train.py

```diff
@@ -0,0 +1,193 @@
+
+    def configure_optimizers(self) -> optim.Optimizer:
+        optimizer = optim.AdamW(self.model.parameters(), lr=self.hparams["lr"], weight_decay=self.hparams["l2"])
+
+        return optimizer
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> float:
+        x = batch["x"]
+
+        batch_size = x.shape[0]
+
+        # sample t
+        t = torch.randint(0, self.timesteps, size=(batch_size,), dtype=torch.long, device=self.device)
+
+        # noise batch
+        x_noisy, added_noise = self.forward_diffusion.forward(x, t)
+
+        # get predicted noise
+        predicted_noise = self.model.forward(x_noisy, t)
+
+        # get loss value for batch
+        loss = self.loss_function_module(predicted_noise, added_noise)
+
+        self.log("loss", loss, on_epoch=True)
+
+        if self.global_step % self.hparams["save_every_n_steps"] == 0:
+            self._save_models()
+
+        return loss
+
+
+def preprocess_dataset(filepath: str, batch_size: int, num_workers: int):
+    dataset = load_dataset(filepath)
+    transformed_dataset = dataset.with_transform(transforms).remove_columns("label")
+
+    dataloader = DataLoader(transformed_dataset["train"], batch_size=batch_size, num_workers=num_workers)
+
+    return dataloader
+
+
+def transforms(examples):
+    transform = Transforms(img_size=28)
+
+    examples["x"] = [transform.img2torch(image.convert("L")) for image in examples["image"]]
+    del examples["image"]
+
+    return examples
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+@hydra.main(config_path="configs", config_name="config-fashion-mnist")
+def train(cfg: Config):
+    # get dataloader
+    dataloader = preprocess_dataset(
+        cfg.hyperparameters.dataset, batch_size=cfg.hyperparameters.batch_size, num_workers=cfg.hyperparameters.num_workers
+    )
+
+    # generate some random id for the run
+    run_id = str(uuid.uuid1())[:8]
+
+    # make directory for logs if doesn't exist
+    makedir_if_not_exists(cfg.paths.log_dir)
+
+    # initializing logger
+    logger = WandbLogger(project="annotated-diffusion", name=f"{cfg.logger.run_name}-{run_id}", save_dir=cfg.paths.log_dir)
+
+    # initializing models
+    model = Unet(
+        dim=cfg.models.unet.dim,
+        dim_mults=cfg.models.unet.dim_mults,
+        in_channels=cfg.models.unet.in_channels,
+        resnet_block_groups=cfg.models.unet.num_resnet_groups,
+    )
+
+    forward_diffusion = ForwardDiffusion(
+        beta_start=cfg.models.forward_diffusion.beta_start,
+        beta_end=cfg.models.forward_diffusion.beta_end,
+        timesteps=cfg.models.forward_diffusion.timesteps,
+        schedule_type=cfg.models.forward_diffusion.schedule_type,
+    )
+
+    # if checkpoint exists load state dict for model and forward diffusion
+    if cfg.paths.load_ckpt_path is not None:
+        checkpoint = torch.load(cfg.paths.load_ckpt_path)
+        model.load_state_dict(checkpoint["model"])
```

Comment:
Popular optimizers also [have state](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html#torch.optim.AdamW.state_dict), and it's a good practice to checkpoint it as well